dev/PushNotifications/PushNotificationChannel.cpp

```diff
@@ -48,18 +51,75 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
     {
-        return m_channel.PushNotificationReceived([weak_self = get_weak(), handler](auto&&, auto&& args)
+        if (IsPackagedApp())
+        {
+            return m_channel.PushNotificationReceived([weak_self = get_weak(), handler](auto&&, auto&& args)
+            {
+                auto strong = weak_self.get();
+                if (strong)
+                {
+                    handler(*strong, winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationReceivedEventArgs>(args));
+                };
+            });
+        }
+        else
         {
-            auto strong = weak_self.get();
-            if (strong)
+            if (!m_isRegisteredWithLRP)
             {
-                handler(*strong, winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationReceivedEventArgs>(args));
-            };
-        });
+                wil::com_ptr<INotificationsLongRunningPlatform> notificationsLongRunningPlatform{
+                    wil::CoCreateInstance<NotificationsLongRunningPlatform, INotificationsLongRunningPlatform>(CLSCTX_LOCAL_SERVER) };
+
+                wchar_t processName[1024];
+                THROW_HR_IF(ERROR_FILE_NOT_FOUND, GetModuleFileNameExW(GetCurrentProcess(), NULL, processName, sizeof(processName) / sizeof(processName[0])) == 0);
+
+                THROW_IF_FAILED(notificationsLongRunningPlatform->RegisterForegroundActivator(this, processName));
+
+                m_isRegisteredWithLRP = true;
+            }
+            return m_foregroundHandlers.add(handler);
+        }
     }
 
     void PushNotificationChannel::PushReceived(winrt::event_token const& token) noexcept
     {
-        m_channel.PushNotificationReceived(token);
+        if (IsPackagedApp())
+        {
+            m_channel.PushNotificationReceived(token);
+        }
+        else
+        {
+            m_foregroundHandlers.remove(token);
+        }
+    }
+
+    HRESULT __stdcall PushNotificationChannel::InvokeAll(ULONG length, byte* start) noexcept try
+    {
+        std::vector<uint8_t> vec;
+
+        for (int i = 0; (ULONG)i < length; i++)
+        {
+            vec.push_back(start[i]);
+        }
+
+        com_array<uint8_t> arr{ start, start + (length * sizeof(uint8_t)) };
+        m_foregroundHandlers(*this, winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationReceivedEventArgs>(start, length));
+        return S_OK;
+    }
+    CATCH_RETURN()
+
+    bool PushNotificationChannel::IsBackgroundTaskBuilderAvailable()
+    {
+        return winrt::Windows::Metadata::ApiInformation::IsMethodPresent(L"Windows.ApplicationModel.Background.BackgroundTaskBuilder", L"SetTaskEntryPointClsid");
+    }
+
+    // Determines if the caller should be treated as packaged app or not.
+    bool PushNotificationChannel::IsPackagedApp()
```

Comment:
Would rename the function to something like: bool IsPackagedAppScenario().

IsPackagedApp gives the wrong impression if the app is packaged but BI isn't supported.